// path-framework/app/path-framework/page/element/button/back-button.component.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Button } from './button.component';
import { Component, Input, Output } from "@angular/core";
var BackButtonComponent = (function () {
    function BackButtonComponent() {
    }
    return BackButtonComponent;
}());
__decorate([
    Input('button'),
    Output('button'),
    __metadata("design:type", BackButton)
], BackButtonComponent.prototype, "button", void 0);
BackButtonComponent = __decorate([
    Component({
        selector: 'path-back-button',
        templateUrl: 'back-button.component.html'
    })
], BackButtonComponent);
export { BackButtonComponent };
var BackButton = (function (_super) {
    __extends(BackButton, _super);
    function BackButton(app, pathService, translationService) {
        return _super.call(this, app, pathService, translationService) || this;
    }
    BackButton.prototype.onClick = function () {
        this.app.navigateBack();
    };
    BackButton.prototype.fromJson = function (modelElement) {
        _super.prototype.fromJson.call(this, modelElement);
        this.type = "backButton";
        this.name = this.translationService.getText("Back");
    };
    return BackButton;
}(Button));
export { BackButton };
