// path-framework/app/path-framework/page/element/button/page-delete-button.component.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Button } from "./button.component";
import { Output, Input, Component } from "@angular/core";
import { KeyUtility } from "../../../key-utility";
var PageDeleteButtonComponent = (function () {
    function PageDeleteButtonComponent() {
    }
    return PageDeleteButtonComponent;
}());
__decorate([
    Input('button'),
    Output('button'),
    __metadata("design:type", PageDeleteButton)
], PageDeleteButtonComponent.prototype, "button", void 0);
PageDeleteButtonComponent = __decorate([
    Component({
        selector: 'path-page-delete-button',
        templateUrl: 'page-delete-button.component.html'
    })
], PageDeleteButtonComponent);
export { PageDeleteButtonComponent };
var PageDeleteButton = (function (_super) {
    __extends(PageDeleteButton, _super);
    function PageDeleteButton(app, pathService, translationService) {
        var _this = _super.call(this, app, pathService, translationService) || this;
        _this.name = translationService.getText("Delete");
        return _this;
    }
    PageDeleteButton.prototype.onClick = function () {
        var _this = this;
        this.getApp().yesNo(this.translationService.getText("DeleteWarningQuestion"), function () {
            _this.pathService.serverDelete(_this.getApp().getBackendUrl(), _this.url, function () {
                _this.getApp().navigateBack();
                _this.getApp().refreshCurrentPage();
            });
        }, function () {
        });
    };
    PageDeleteButton.prototype.fromJson = function (modelElement) {
        _super.prototype.fromJson.call(this, modelElement);
        this.type = "deleteButton";
        this.url = KeyUtility.translateUrl(modelElement["url"], null, false, this.parentPageElement);
    };
    return PageDeleteButton;
}(Button));
export { PageDeleteButton };
