// path-framework/app/path-framework/form/field/checkbox/checkbox-group.component.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input, Output } from '@angular/core';
import { ValueField } from "../value-field";
var CheckboxGroupComponent = (function () {
    function CheckboxGroupComponent() {
    }
    return CheckboxGroupComponent;
}());
__decorate([
    Input('field'),
    Output('field'),
    __metadata("design:type", CheckboxGroupField)
], CheckboxGroupComponent.prototype, "field", void 0);
CheckboxGroupComponent = __decorate([
    Component({
        selector: 'path-checkbox-group',
        templateUrl: 'checkbox-group.component.html'
    })
], CheckboxGroupComponent);
export { CheckboxGroupComponent };
var CheckboxGroupField = (function (_super) {
    __extends(CheckboxGroupField, _super);
    function CheckboxGroupField() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._entries = [];
        return _this;
    }
    Object.defineProperty(CheckboxGroupField.prototype, "entries", {
        get: function () {
            return this._entries;
        },
        set: function (value) {
            this._entries = value;
        },
        enumerable: true,
        configurable: true
    });
    CheckboxGroupField.prototype.doClick = function (key) {
        var values = [];
        for (var _i = 0, _a = this.entries; _i < _a.length; _i++) {
            var entry = _a[_i];
            if (entry.selected) {
                values.push(entry.key);
            }
        }
        this.setValue(values);
    };
    CheckboxGroupField.prototype.setValue = function (values) {
        _super.prototype.setValue.call(this, values);
        for (var _i = 0, _a = this.entries; _i < _a.length; _i++) {
            var entry = _a[_i];
            entry.selected = false;
        }
        for (var _b = 0, values_1 = values; _b < values_1.length; _b++) {
            var value = values_1[_b];
            for (var _c = 0, _d = this.entries; _c < _d.length; _c++) {
                var entry = _d[_c];
                if (entry.key == value) {
                    entry.selected = true;
                }
            }
        }
    };
    CheckboxGroupField.prototype.fromJson = function (modelFormField) {
        _super.prototype.fromJson.call(this, modelFormField);
        for (var _i = 0, _a = modelFormField["data"]; _i < _a.length; _i++) {
            var entryModel = _a[_i];
            var entry = new CheckboxGroupEntry();
            entry.name = entryModel.name;
            entry.key = entryModel.key;
            entry.selected = false;
            this.entries.push(entry);
        }
    };
    return CheckboxGroupField;
}(ValueField));
export { CheckboxGroupField };
var CheckboxGroupEntry = (function () {
    function CheckboxGroupEntry() {
    }
    Object.defineProperty(CheckboxGroupEntry.prototype, "key", {
        get: function () {
            return this._key;
        },
        set: function (value) {
            this._key = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckboxGroupEntry.prototype, "name", {
        get: function () {
            return this._name;
        },
        set: function (value) {
            this._name = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CheckboxGroupEntry.prototype, "selected", {
        get: function () {
            return this._selected;
        },
        set: function (value) {
            this._selected = value;
        },
        enumerable: true,
        configurable: true
    });
    return CheckboxGroupEntry;
}());
export { CheckboxGroupEntry };
